### DISCUSSION TITLE
C-h key bindings like Emacs

### DISCUSSION THREAD

**Discussion by g-gundam** [lem-project/lem#1500] [Show and tell]
Coming from Emacs, I expected to find some bindings off of <kbd>C-h</kbd> that would help me introspect Lem.  By scanning through <kbd>M-x</kbd>'s completions, I saw that similar functions existed, but they weren't bound to anything as far as I could tell.  If you were looking for some familiar keybindings off of <kbd>C-h</kbd> like I was, try putting this in your `~/.lem/init.lisp`.

```lisp
(define-keys *global-keymap*
  ("C-h a" 'lem-lisp-mode/internal:lisp-apropos)
  ("C-h b" 'lem-core/commands/help::describe-bindings)
  ("C-h f" 'lem-lisp-mode/internal:lisp-describe-symbol)
  ("C-h k" 'lem-core/commands/help::describe-key)
  ("C-h m" 'lem-core/commands/help::describe-mode))
```

It's not exactly the same as Emacs, but it's close enough in spirit.  I think this will help people coming from Emacs learn how to use Lem.

## Resources
- [Configuration](https://lem-project.github.io/usage/configuration/)
- [Common Lisp Commands](https://lem-project.github.io/usage/common_lisp/)

**Comment by LeeLaffan** [comment]
Hi, this is really useful to have, thanks for sharing.

I seen when browsing the docs, it looks like there should be keybindings for these: [Docs](https://github.com/lem-project/lem/blob/4b3fa46b0dd849a60b336fe83781ecb2ec81cd49/docs/default-keybindings.md?plain=1#L172)

Have you considered making a pull request to add these? I think for users coming from emacs or vim(need to know editor commands matching vim commands) this small addition would be a big change to how nice to initial experience making a config is. In that case I would argue for your C-h keybinds, but it could be discussed.

I think a good place to add the describe ones(could add the other appropriately) would be: [help.lisp](https://github.com/lem-project/lem/blob/4b3fa46b0dd849a60b336fe83781ecb2ec81cd49/src/commands/help.lisp#L12)

Thanks again :)

**Reply by g-gundam** [reply]
I didn't know about those docs, but it sure looks like someone was getting ready to bind them to something.  Who knows why they were left unbound.  I personally think the Emacs bindings would be fine here, but they did deviate from Emacs by putting describe-key on `C-x ?` instead of `C-h k`.

I can send them a pull request, but I don't know if it'll get any traction.

**Reply by g-gundam** [reply]
This is turning out to be trickier than I anticipated.  There are a couple of problems.

## lem on the terminal treats `C-h` like a backspace
- I managed to create some `C-h` bindings, but I had to sacrifice backspace to get them.
- That must be why `describe-key` was on `C-x ?`.

## I don't know how to access the `lem-lisp-mode` functions.
- Everything that's under `lem-core/commands/help` was in scope, so those were no problem to bind.
- However, my Common Lisp understanding is weak, and I don't know how to make `lem-lisp-mode/internal` available to me while in help.lisp.

I guess I won't be sending a pull request after all.  Sorry man.  I tried.

**Reply by LeeLaffan** [reply]
Hi, oh wow, always the simple things that turn out to be the most complicated!

I had a mess around with it(thought i had it but it was your config in mine lol), no idea, hopefully someone more experienced can tell us. If it is a package loading order problem then I think it's best to just define lisp-describe-symbol in lem-lisp-mode/internal, if that's not the best thing anyways.

As for the terminal stuff, that sounds... messy. How about `C-x h _`? And in your pull request you can explain why C-h doesn't work. I don't think you wasted anytime if you learned any shared these things as you have :)

**Reply by g-gundam** [reply]
I want to backtrack on the backspace problem a little bit.  I happened to be using `cool-retro-term` when I was having the `C-h` problem.  However, in `wezterm`, it's behaving better.

If I run emacs under `cool-retro-term`, it has the same `C-h` trouble, so it's likely a bug in `cool-retro-term` than a bug in lem or emacs.

**Reply by LeeLaffan** [reply]
Oh nice. And then the functions work if you put them in the matching package. Likely a load order issue with extensions vs the base libraries, might make sense you can't, I don't know.

That's good because the main thing was matching emacs, `C-x h` wasn't as good. I think you should try the PR personally and they can tell you what can be done differently if anything, right?

At the very least your original post keybinds should be on the wiki, I would imagine a lot of people using Lem will try to use these keys early on(I did!) and it can be hard to tell what you are not accessing vs what is not implemented.

Your call, thanks for the config either way :)

**Comment by g-gundam** [comment]
I don't know if this is recent but check this out:
![image](https://github.com/user-attachments/assets/9d926f98-f65a-4a2f-bd75-f8d69672a846)

That's from the [Binding Keys](https://lem-project.github.io/usage/configuration/#binding-keys) section.  Was that there before?  I still wonder why they deviate from emacs here though.  They really don't want to bind `C-h`.

**Reply by vindarel** [reply]
oh my you are right^^ I updated the page.

**Reply by g-gundam** [reply]
Thanks for updating the docs.
